File: eFormAPI/Plugins/TimePlanning.Pn/TimePlanning.Pn/Infrastructure/Models/Planning/ReconcileThroughResultModel.cs
#nullable enable
namespace TimePlanning.Pn.Infrastructure.Models.Planning;

using System;
using System.Collections.Generic;

public class ReconcileThroughResultModel
{
    /// <summary>Where each worker's boundary landed. Per worker, not shared:
    /// the mark falls on that worker's latest day with a registration at or
    /// before the requested date, so a staircase has no single landing date.</summary>
    public Dictionary<int, DateTime> LandedOnBySiteId { get; set; } = new();

    /// <summary>Workers whose boundary actually moved. Excludes no-ops.</summary>
    public int Applied { get; set; }

    /// <summary>Already reconciled at or past the target. Moving them back would
    /// be an unlock, which is deliberately a separate, heavier action.</summary>
    public List<int> SkippedAlreadyFurtherForward { get; set; } = new();

    /// <summary>No registration at or before the target, so there was nothing
    /// to mark. A distinct case from the above — the spec distinguishes them.</summary>
    public List<int> SkippedNoRegistration { get; set; } = new();

    /// <summary>Already marked on exactly the landing day; nothing changed.</summary>
    public List<int> AlreadyReconciledSiteIds { get; set; } = new();
}

File: eform-client/src/app/plugins/modules/time-planning-pn/models/plannings/reconcile-through-result.model.ts
/**
 * Mirrors the C# ReconcileThroughResultModel (plan Task 4) field for field.
 *
 * Per-worker landing, because the boundary is a staircase: the mark falls on each
 * worker's own latest registered day at or before the requested date, so one shared
 * "landedOn" would name the wrong date for most of them.
 */
export class ReconcileThroughResultModel {
  /** siteId -> the day the mark landed on. JSON object keys arrive as strings. */
  landedOnBySiteId: { [siteId: number]: string } = {};
  /** Workers whose boundary actually moved. Excludes no-ops. */
  applied: number;
  /** Already reconciled at or past the target. Moving them back would be an unlock. */
  skippedAlreadyFurtherForward: number[] = [];
  /** No registration at or before the target, so there was nothing to mark. */
  skippedNoRegistration: number[] = [];
  /** Already marked on exactly the landing day; nothing changed. */
  alreadyReconciledSiteIds: number[] = [];
}
